consensus/clique/src/main/java/tech/pegasys/pantheon/consensus/clique/CliqueExtraData.java

```diff
@@ -47,9 +57,28 @@ public CliqueExtraData(
     this.vanityData = vanityData;
     this.proposerSeal = Optional.ofNullable(proposerSeal);
     this.validators = validators;
+    proposerAddress =
+        Suppliers.memoize(() -> CliqueBlockHashing.recoverProposerAddress(header, this));
```

Comment:
When `header` is `null`, looks like `getProposerAddress()` will throw an error

Reply:
So all these `null`s are required because when you're mining a block you need to have something for the extra data before you can calculate the hash and sign it. Except for one slightly contrived case in test setup, whenever header is `null`, proposerSeal is also `null` or empty - so in either case you can't recover the proposer address because there isn't one.

But I like the idea of a static encode function and it fits in quite well so have done that.  You can still wind up with a `CliqueExtraData` instance with an invalid principal seal but we need to handle that because that's what the genesis block has.